feat(parser): add a common SARIF parser and use it for checkstyle

kmoschcau opened the pull request:
This adds a [SARIF](https://sarifweb.azurewebsites.net/) parser and uses it for
checkstyle.

kmoschcau commented:
I just realized, that `for_sarif` likely needs some documentation in the README. I can add that tomorrow. Also I'm not sure where to put the spec for this parsing function. I didn't get the test runner working so far either. Any help with that would be appreciated.

mfussenegger commented:
>  Also I'm not sure where to put the spec for this parsing function

I'd keep the checkstyle spec but adapt the input to the sarif output.

> . I didn't get the test runner working so far either. Any help with that would be appreciated.

See https://github.com/nvim-lua/nvim-lua-plugin-template?tab=readme-ov-file#run-tests

To help further I'd need to know exact errors.

mfussenegger commented on lua/lint/parser.lua:
```diff
@@ -11,6 +11,75 @@ local severity_by_qftype = {
+        local file_bufnr =
+          vim.uri_to_bufnr(vim.uri_from_fname(vim.fs.abspath(location.physicalLocation.artifactLocation.uri)))
+
+        -- TODO: This check can be removed, once nvim-lint supports multiple
+        -- buffers.
+        -- https://github.com/mfussenegger/nvim-lint/issues/716
```
```suggestion
```

If multi-buffer/project linting ever gets supported I'll likely need to go through the various linters anyway, no extra TODO needed.

mfussenegger commented on lua/lint/parser.lua:
```diff
@@ -11,6 +11,75 @@ local severity_by_qftype = {
+
+    local decoded = vim.json.decode(output) or {}
+
+    local run = decoded.runs and decoded.runs[1] or {}
```
I'm not very familiar with sarif but based on the [runs spec](https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/sarif-v2.1.0-errata01-os-complete.html#_Toc141790732) I get the impression that it might be more appropriate to loop over all runs and collect diagnostics for all of them?

kmoschcau commented:
OK I got the tests working on a Linux machine. Something tries to run `~` as an executable on Windows, which of course fails.

kmoschcau commented:
I addressed all comments and added a comprehensive spec file for the `for_sarif` parser. This includes the `checkstyle` tests and adds some additional ones.

kmoschcau commented:
I found a bug with checkstyle on Windows, when using the SARIF format: https://github.com/checkstyle/checkstyle/issues/16127
Up to you, if you want to hold this PR back until this is solved.

kmoschcau commented:
The pull request in checkstyle was merged. I don't know when they will release it, but at least that's not a blocker anymore.

kmoschcau commented:
@mfussenegger The fix is in checkstyle 10.21.3, which was release 3 weeks ago.
